File: 0134-gas-station/0134-gas-station.py
class Solution:
    def bruteForce(self, gas: List[int], cost: List[int]) -> int:
        
        N = len(gas)
        
        for i in range(N):
            if gas[i] < cost[i]:
                continue
            
            curGas = gas[i]
            j = (i + 1) % N
            
            costForNextMove = cost[i]
            gasEarnFromNextStation = gas[j]
            
            curGas = curGas - costForNextMove + gasEarnFromNextStation
            
            while i!=j:
                if curGas < cost[j]:
                    break
                
                costForNextMove = cost[j]
                
                j = (j + 1)%N
                
                gasEarnFromNextStation = gas[j]
                
                curGas = curGas - costForNextMove + gasEarnFromNextStation
            
            if i==j:
                return i
        
        return -1
           
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        
        totalGasNeeded = sum(gas)
        totalCostNeeded = sum(cost)
        
        if totalGasNeeded < totalCostNeeded:
            return -1
        
        N = len(gas)
        result_idx = 0 # probable answer
        totalCost = 0
        
        for i in range(N):
            totalCost += gas[i] - cost[i]
            
            if totalCost < 0:
                totalCost = 0
                result_idx = i+1
        
        return result_idx
    
        
    

File: 0134-gas-station/0134-gas-station.cpp
class Solution {
private:
    int bruteForce(vector<int>& gas, vector<int>& cost){
        
        int n = gas.size();
        
        for(int i=0; i<n; i++){
            
            if(gas[i] < cost[i])
                continue;
            
            int costForMovingToNextStation = cost[i];
            
            int j = (i + 1) % n;
            int gasEarnFromNextStationJ = gas[j];
            
            int curGas = gas[i] - costForMovingToNextStation + gasEarnFromNextStationJ;
            
            while(j != i){
                
                if(curGas < cost[j]) break;
                
                costForMovingToNextStation = cost[j];
                j = (j + 1) % n;
                
                gasEarnFromNextStationJ = gas[j];
                
                curGas = curGas - costForMovingToNextStation + gasEarnFromNextStationJ;
            }

            if(i == j) return i;
        }
        
        return -1;
    }
    
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        
        int n = gas.size();
        
        int totalGas = 0;
        int totalCost = 0;
        
        for(int i=0; i<n; i++){
            totalGas += gas[i];
            totalCost += cost[i];
        }
        
        if(totalGas < totalCost) return -1;
        
        int result = 0;
        int total=0;
        
        for(int i=0; i<n; i++){
            total += gas[i] - cost[i];
            if(total < 0) {
                total = 0;
                result = i+1;
            }
        }
        
        return result;
    }
};